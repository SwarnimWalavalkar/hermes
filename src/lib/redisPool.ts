import {
  Factory,
  Pool,
  createPool,
  Options as PoolOptions,
} from "generic-pool";
import Redis, { Redis as IRedis, RedisOptions } from "ioredis";
import { EventEmitter } from "stream";

export class RedisPool extends EventEmitter {
  private redisOpts: RedisOptions;
  private poolOpts: PoolOptions;
  private pool: Pool<IRedis>;

  constructor(redisOpts: RedisOptions, poolOpts: PoolOptions) {
    super();

    this.redisOpts = redisOpts;
    this.poolOpts = poolOpts;

    this.pool = this.constructPool();
  }

  private constructPool(): Pool<IRedis> {
    const factory: Factory<IRedis> = {
      create: (): Promise<IRedis> => {
        const context = this;
        return new Promise((resolve, reject) => {
          const client: IRedis = new Redis(this.redisOpts);

          client
            .on("error", (e: Error) => {
              context.emit("error", e, client);
              reject();
            })
            .on("connect", () => {
              context.emit("connect", client);
            })
            .on("ready", () => {
              context.emit("ready", client);
              resolve(client);
            })
            .on("reconnecting", () => {
              context.emit("reconnecting", client);
            });
        });
      },

      destroy: (client: IRedis): Promise<void> => {
        const context = this;
        return new Promise((resolve) => {
          client
            .on("close", (e: Error) => {
              context.emit("close", e, client);
            })
            .on("end", () => {
              context.emit("disconnected", client);
              resolve();
            })
            .disconnect();
        });
      },

      validate: (client): Promise<boolean> => {
        return new Promise((resolve) => {
          if (
            client.status === "connecting" ||
            client.status === "connect" ||
            client.status === "ready"
          ) {
            resolve(true);
          } else {
            resolve(false);
          }
        });
      },
    };

    return createPool(factory, this.poolOpts);
  }

  getConnection(priority?: number) {
    return this.pool.acquire(priority);
  }

  release(client: IRedis) {
    return this.pool.release(client);
  }

  disconnect(client: IRedis) {
    return this.pool.destroy(client);
  }

  async end() {
    await this.pool.drain();

    const res = await this.pool.clear();
    this.emit("end");
    return res;
  }

  async execute<T>(fn: (client: IRedis) => Promise<T>, priority?: number) {
    const client = await this.pool.acquire(priority);
    const result = await fn(client);
    await this.release(client);
    return result;
  }
}

export { Options as PoolOptions } from "generic-pool";
