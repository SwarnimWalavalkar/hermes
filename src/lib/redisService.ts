import Redis, { RedisOptions } from "ioredis";

export interface RedisService {
  connect(): Promise<void>;
  disconnect(): Promise<void>;
  addToStream(
    streamName: string,
    id: string,
    ...args: Array<string>
  ): Promise<void>;
  ackMessages(
    streamName: string,
    groupName: string,
    ...messageIds: Array<string>
  ): Promise<void>;
  createConsumerGroup(streamName: string, groupName: string): Promise<void>;
  readStreamAsConsumerGroup(
    streamName: string,
    count?: number,
    blockMs?: number,
    group?: string,
    consumer?: string
  ): Promise<Array<string> | null>;
  autoClaimMessages(
    streamName: string,
    count?: number,
    group?: string,
    consumer?: string,
    minIdleTime?: number
  ): Promise<Array<string> | null>;
}

export function RedisService(
  redisOptions: RedisOptions,
  groupName: string,
  consumerName: string
): RedisService {
  let subscriber: Redis;
  let publisher: Redis;

  async function connect() {
    subscriber = new Redis({ ...redisOptions });
    publisher = new Redis({ ...redisOptions });
  }

  async function disconnect() {
    await deleteConsumer();

    subscriber.disconnect();
    publisher.disconnect();
  }

  async function deleteConsumer() {
    try {
      const key = `${redisOptions.keyPrefix}*`;

      const keys = await subscriber.keys(key);

      if (keys.length) {
        for (const key of keys) {
          const groups: string[] = (await subscriber.xinfo(
            "GROUPS",
            key
          )) as string[];

          if (groups.length)
            await subscriber.xgroup(
              "DELCONSUMER",
              key,
              groupName,
              consumerName
            );
        }
      }
    } catch (error: any) {
      console.error("[HERMES] Error deleting consumer");
      throw error;
    }
  }

  async function addToStream(
    streamName: string,
    id: string,
    ...args: Array<string>
  ) {
    try {
      await publisher.xadd(streamName, id, ...args);
    } catch (error: any) {
      console.error(`[HERMES] Error adding to stream: ${error.message}`);
      throw error;
    }
  }

  async function ackMessages(
    streamName: string,
    groupName: string,
    ...messageIds: Array<string>
  ) {
    try {
      await subscriber.xack(streamName, groupName, ...messageIds);
    } catch (error: any) {
      console.error(`[HERMES] Error acknowledging messages: ${error.message}`);
      throw error;
    }
  }

  async function createConsumerGroup(streamName: string, groupName: string) {
    try {
      await subscriber.xgroup(
        "CREATE",
        `${redisOptions.keyPrefix}${streamName}`,
        groupName,
        "0",
        "MKSTREAM"
      );
    } catch (error: any) {
      if (error.message.includes("BUSYGROUP")) {
        return;
      }
      console.error(
        `[HERMES] Error while creating consumer group: ${error.message}`
      );
      throw error;
    }
  }

  async function readStreamAsConsumerGroup(
    streamName: string,
    count: number = 1,
    blockMs: number = 1,
    group: string = groupName,
    consumer: string = consumerName
  ): Promise<Array<string> | null> {
    try {
      const results: string[][] = (await subscriber.xreadgroup(
        "GROUP",
        group,
        consumer,
        "COUNT",
        count,
        "BLOCK",
        blockMs,
        "STREAMS",
        streamName,
        ">"
      )) as string[][];

      if (results && results.length && results[0]) {
        const [_key, messages] = results[0];

        if (messages) return messages as unknown as string[];
      }
      return null;
    } catch (error: any) {
      if (error.message.includes("NOGROUP")) {
        console.log(`${error.message} ...CREATING GROUP`);
        await createConsumerGroup(streamName, group);
        return null;
      }
      console.error(`[HERMES] Error reading stream: ${error.message}`);
      throw error;
    }
  }

  async function autoClaimMessages(
    streamName: string,
    count: number = 1,
    group: string = groupName,
    consumer: string = consumerName,
    minIdleTime: number = 5000
  ): Promise<Array<string> | null> {
    try {
      const results: string[][] = (await subscriber.xautoclaim(
        streamName,
        group,
        consumer,
        minIdleTime,
        "0-0",
        "COUNT",
        count
      )) as string[][];

      if (results && results.length && results[0]) {
        const [_key, messages] = results[0];

        if (messages) return messages as unknown as string[];
      }
      return null;
    } catch (error: any) {
      console.error(`[HERMES] Error auto claiming messages: ${error.message}`);
      throw error;
    }
  }

  return {
    connect,
    disconnect,
    addToStream,
    ackMessages,
    autoClaimMessages,
    createConsumerGroup,
    readStreamAsConsumerGroup,
  };
}
