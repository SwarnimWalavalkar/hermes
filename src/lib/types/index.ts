import z from "zod";

export type Maybe<T> = T | null | undefined;

export interface IBus {
  subscribe<T>(
    topic: string,
    callback: (msgData: { data: T; msgId: string }) => Promise<void>
  ): void | Promise<void>;
  publish<T>(topic: string, data: T): Promise<void>;
}

export interface RedisMessage<T> {
  data: T;
  maxRetries: number;
  retryCount: number;
}

export interface IMsg {
  id: string;
  retryCount: number;
  maxRetries: number;
  ack: () => Promise<void>;
}

export interface IEvent<MessagePayload> {
  subscribe(
    fn: (msgData: { data: MessagePayload; msg: IMsg }) => void | Promise<void>
  ): Promise<void>;
  publish(
    data: MessagePayload,
    options?: { maxRetries?: number }
  ): Promise<void>;
}

export interface IService<RequestType, ResponseType> {
  reply(
    fn: (msgData: {
      reqData: RequestType;
      msgId: string;
    }) => ResponseType | Promise<ResponseType>
  ): void | Promise<void>;
  request(reqData: RequestType): Promise<ResponseType>;
}

export interface IHermes {
  connect(): Promise<IHermes>;
  disconnect(): Promise<void>;
  registerEvent<MessagePayload>(
    topic: string,
    payloadSchema: z.Schema<MessagePayload>,
    options?: { maxRetries?: number }
  ): Promise<IEvent<MessagePayload>>;
  registerService<RequestType, ResponseType>(
    topic: string,
    requestSchema: z.Schema<RequestType>,
    responseSchema: z.Schema<ResponseType>
  ): Promise<IService<RequestType, ResponseType>>;
}
