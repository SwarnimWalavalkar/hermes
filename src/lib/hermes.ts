import { RedisService } from "./redis/redisService";
import { RedisOptions } from "ioredis";
import { Redis as UpstashRedis } from "@upstash/redis";
import { randomBytes } from "crypto";
import sleep from "../utils/sleep";
import { z } from "zod";
import { PoolOptions } from "./redis/redisPool";
import { UpstashRedisService } from "./redis/upstashRedis";

type Maybe<T> = T | null | undefined;

interface IBus {
  subscribe<T>(
    topic: string,
    callback: (msgData: { data: T; msgId: string }) => Promise<void>
  ): void | Promise<void>;
  publish<T>(topic: string, data: T): Promise<void>;
}

export interface IMsg {
  id: string;
  ack: () => Promise<void>;
}

export interface IEvent<MessagePayload> {
  subscribe(
    fn: (msgData: { data: MessagePayload; msg: IMsg }) => void | Promise<void>
  ): Promise<void>;
  publish(data: MessagePayload): Promise<void>;
}

export interface IService<RequestType, ResponseType> {
  reply(
    fn: (msgData: {
      reqData: RequestType;
      msgId: string;
    }) => ResponseType | Promise<ResponseType>
  ): void | Promise<void>;
  request(reqData: RequestType): Promise<ResponseType>;
}

export type HermesConfig = {
  durableName: string;
} & (
  | {
      persistanceType: "REDIS";
      redisOptions: RedisOptions;
      poolOptions?: PoolOptions;
    }
  | { persistanceType: "UPSTASH"; upstashRedis: UpstashRedis }
);

export interface IHermes {
  connect(): Promise<IHermes>;
  disconnect(): Promise<void>;
  registerEvent<MessagePayload>(
    topic: string,
    payloadSchema: z.Schema<MessagePayload>
  ): Promise<IEvent<MessagePayload>>;
  registerService<RequestType, ResponseType>(
    topic: string,
    requestSchema: z.Schema<RequestType>,
    responseSchema: z.Schema<ResponseType>
  ): Promise<IService<RequestType, ResponseType>>;
}

const KEY_PREFIX = "hermes:";
const DEFAULT_POOL_OPTIONS = { min: 0, max: 20 };

export function Hermes(config: HermesConfig): IHermes {
  let redisService: RedisService | UpstashRedisService;
  let isAlive = false;

  const consumerName = randomBytes(16).toString("hex");
  const groupName = config.durableName;

  async function connect() {
    if (config.persistanceType === "REDIS") {
      if (!config.poolOptions) {
        config.poolOptions = DEFAULT_POOL_OPTIONS;
      }

      redisService = RedisService(
        {
          ...config.redisOptions,
          keyPrefix: KEY_PREFIX,
          poolOptions: config.poolOptions,
        },
        groupName,
        consumerName
      );

      await redisService.connect();
    }
    if (config.persistanceType === "UPSTASH") {
      redisService = UpstashRedisService(
        config.upstashRedis,
        { keyPrefix: KEY_PREFIX },
        groupName,
        consumerName
      );
    }

    isAlive = true;

    // @ts-expect-error ignore
    return this;
  }

  async function disconnect() {
    isAlive = false;

    await sleep(100);

    if (config.persistanceType === "REDIS") {
      // @ts-expect-error ignore
      await redisService.disconnect();
    } else if (config.persistanceType === "UPSTASH") {
      // @ts-expect-error ignore
      await redisService.cleanup();
    }
  }

  async function* getStreamMessageGenerator(streamName: string, count: number) {
    let fetchNewMessages = true;
    while (isAlive) {
      const results = fetchNewMessages
        ? await redisService.readStreamAsConsumerGroup(streamName, count)
        : await redisService.autoClaimMessages(streamName, count);

      fetchNewMessages = !fetchNewMessages;

      if (!results || !results.length) {
        continue;
      }

      for (const message of results) {
        yield message;
      }
    }
  }

  async function registerEvent<MessagePayload>(
    topic: string,
    payloadSchema: z.Schema<MessagePayload>
  ): Promise<IEvent<MessagePayload>> {
    async function subscribe(
      callback: (msgData: {
        data: MessagePayload;
        msg: { id: string; ack: () => Promise<void> };
      }) => Promise<void>
    ) {
      try {
        const generator = getStreamMessageGenerator(topic, 10);

        for await (const message of generator) {
          if (message.length && message[1] && message[1][1]) {
            const data: Maybe<MessagePayload> = JSON.parse(message[1][1]);
            const msgId: string = String(message[0]);

            let parsedData: MessagePayload;
            try {
              parsedData = payloadSchema.parse(data);
            } catch (error) {
              console.error(
                `[HERMES] Message Bus: ${topic}, message parse error`
              );
              throw new Error("Message Parse Error");
            }

            await callback({
              data: parsedData,
              msg: {
                id: msgId,
                ack: () => redisService.ackMessages(topic, groupName, msgId),
              },
            });
          }
        }
      } catch (error) {
        console.error("[HERMES] Error while subscribing:", error);
        throw error;
      }
    }

    async function publish(payload: MessagePayload): Promise<void> {
      let parsedData: MessagePayload;
      try {
        parsedData = payloadSchema.parse(payload);
      } catch (error) {
        console.error(`[HERMES] Message Bus: ${topic}, message parse error`);
        throw new Error("Message Parse Error");
      }
      await redisService.addToStream(
        topic,
        "*",
        "data",
        JSON.stringify(parsedData)
      );
    }

    return { subscribe, publish };
  }

  const bus: IBus = {
    async subscribe<T>(
      topic: string,
      callback: (msgData: { data: T; msgId: string }) => Promise<void>
    ): Promise<void> {
      try {
        const generator = getStreamMessageGenerator(topic, 10);

        for await (const message of generator) {
          if (message.length && message[1] && message[1][1]) {
            const data: T = JSON.parse(message[1][1]);
            const msgId: string = String(message[0]);

            await redisService.ackMessages(topic, groupName, msgId);

            await callback({ data, msgId });
          }
        }
      } catch (error) {
        console.error("[HERMES] Error while subscribing:", error);
        throw error;
      }
    },
    async publish<T>(topic: string, data: T): Promise<void> {
      await redisService.addToStream(topic, "*", "data", JSON.stringify(data));
    },
  };

  async function registerService<RequestType, ResponseType>(
    topic: string,
    requestSchema: z.ZodTypeAny,
    responseSchema: z.ZodTypeAny
  ) {
    async function reply(
      callback: (msgData: {
        reqData: RequestType;
        msgId: string;
      }) => ResponseType | Promise<ResponseType>
    ): Promise<void> {
      await bus.subscribe<RequestType>(topic, async ({ msgId, data }) => {
        let parsedReqData: RequestType;

        try {
          parsedReqData = requestSchema.parse(data);
        } catch (error) {
          console.error(`[HERMES] Service: ${topic}, request parse error`);
          throw new Error("Request Parse Error");
        }

        const res = await callback({ reqData: parsedReqData, msgId });

        let parsedResponseData: ResponseType;

        try {
          parsedResponseData = responseSchema.parse(res);
        } catch (error) {
          console.error(`[HERMES] Service: ${topic}, reply parse error`);
          throw new Error("Reply Parse Error");
        }

        await redisService.ackMessages(topic, groupName, msgId);

        await redisService.addToStream(
          `${topic}-res`,
          "*",
          "data",
          JSON.stringify(parsedResponseData),
          "reqMsgId",
          msgId
        );
      });
    }

    async function request(reqData: RequestType): Promise<ResponseType> {
      let parsedRequestData: RequestType;

      try {
        parsedRequestData = requestSchema.parse(reqData);
      } catch (error) {
        console.error(`[HERMES] Service: ${topic}, request parse error`);
        throw new Error("Request Parse Error");
      }

      const responseTopic = `${topic}-res`;

      await bus.publish<RequestType>(topic, parsedRequestData);

      const responseGenerator = getStreamMessageGenerator(responseTopic, 1);

      const messageResp = await responseGenerator.next();

      if (!messageResp.done && messageResp.value) {
        const message = messageResp.value;

        if (message[0] && message[1] && message[1][1]) {
          const data: Maybe<ResponseType> = JSON.parse(message[1][1]);
          const msgId = message[0];

          let parsedData: ResponseType;

          try {
            parsedData = responseSchema.parse(data);
          } catch (error) {
            console.error(`[HERMES] Service: ${topic}, reply parse error`);
            throw new Error("Reply Parse Error");
          }

          await redisService.ackMessages(responseTopic, groupName, msgId);

          return parsedData;
        } else {
          console.error("[HERMES] Invalid redis payload");
          throw Error("Invalid redis payload");
        }
      } else {
        console.error("[HERMES] Unexpected Error while making a request");
        throw Error("Unexpected Error while making a request");
      }
    }

    return {
      request,
      reply,
    };
  }

  return {
    connect,
    disconnect,
    registerEvent,
    registerService,
  };
}
