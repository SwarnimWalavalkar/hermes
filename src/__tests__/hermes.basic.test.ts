import {
  afterAll,
  expect,
  describe,
  it,
  beforeAll,
  expectTypeOf,
  vi,
} from "vitest";
import { Hermes } from "..";
import { z } from "zod";
import { IHermes, IMsg } from "../lib/types";
import Redis from "ioredis";

const redisConfig = {
  host: process.env.REDIS_HOST || "0.0.0.0",
  port: Number(process.env.REDIS_PORT) || 6379,
  password: process.env.REDIS_PASSWORD || "",
};

const testRedis = new Redis({ ...redisConfig });

describe("Initialize", async () => {
  it("should establish a connection and initialize successfully", async () => {
    const hermes = await Hermes({
      durableName: "hermesTest",
      redisOptions: { ...redisConfig },
      poolOptions: { min: 0, max: 20 },
    }).connect();

    await hermes.disconnect();

    expect(hermes).toBeDefined();
  });
});

describe("Message Bus", async () => {
  let hermes: IHermes;
  beforeAll(async () => {
    hermes = await Hermes({
      durableName: "hermesTest",
      redisOptions: { ...redisConfig },
      poolOptions: { min: 0, max: 20 },
    }).connect();
  });

  it("should publish and subscribe to a topic", async () => {
    const topic = "test-topic";

    const payloadSchema = z.object({ message: z.string() });
    const msgPayload: z.infer<typeof payloadSchema> = { message: "hello" };

    const event = await hermes.registerEvent(topic, payloadSchema);
    const eventCallback = {
      fn: async ({
        msg,
        data,
      }: {
        msg: IMsg;
        data: z.infer<typeof payloadSchema>;
      }) => {
        console.log("Message received:", data);

        await msg.ack();
      },
    };

    const callbackFnSpy = vi.spyOn(eventCallback, "fn");

    event.subscribe(eventCallback.fn);
    await event.publish(msgPayload);

    await new Promise((resolve) => setTimeout(resolve, 24));
    expect(callbackFnSpy).toHaveBeenCalledOnce();
  });

  it(
    "should retry with exponential backoff on error",
    { timeout: 7500 },
    async () => {
      const topic = "fail-test-topic";

      const payloadSchema = z.object({ message: z.string() });
      const msgPayload: z.infer<typeof payloadSchema> = { message: "hello" };

      const event = await hermes.registerEvent(topic, payloadSchema);
      const eventCallback = {
        fn: async ({
          msg,
          data,
        }: {
          msg: IMsg;
          data: z.infer<typeof payloadSchema>;
        }) => {
          console.log("MSG_MAX_RETRIES", msg.maxRetries);
          console.log("MSG_RETRY_COUNT", msg.retryCount);

          if (msg.retryCount === msg.maxRetries - 1) {
            console.log("SUCCESS");

            await msg.ack();
          } else {
            console.log("FAIL");
            throw new Error("Test Error");
          }
        },
      };

      const callbackFnSpy = vi.spyOn(eventCallback, "fn");

      event.subscribe(eventCallback.fn);
      await event.publish(msgPayload);

      await new Promise((resolve) => setTimeout(resolve, 7000));
      expect(callbackFnSpy).toHaveBeenCalledTimes(3);
    }
  );

  it("should go to the dlq on final failure", { timeout: 15000 }, async () => {
    const topic = "dlq-test-topic";

    const payloadSchema = z.object({ message: z.string() });
    const msgPayload: z.infer<typeof payloadSchema> = { message: "hello" };

    const event = await hermes.registerEvent(topic, payloadSchema);
    const eventCallback = {
      fn: async () => {
        throw new Error("DQL Test Error");
      },
    };

    const callbackFnSpy = vi.spyOn(eventCallback, "fn");

    event.subscribe(eventCallback.fn);
    await event.publish(msgPayload);

    await new Promise((resolve) => setTimeout(resolve, 14500));

    expect(callbackFnSpy).toHaveBeenCalledTimes(4);

    const DQLKey = `hermes:${topic}-dlq`;
    const dlqSize = await testRedis.zcard(DQLKey);

    expect(dlqSize).toBeGreaterThanOrEqual(1);
  });

  afterAll(async () => {
    await hermes.disconnect();
  });
});

describe("Service", async () => {
  let hermes: IHermes;
  beforeAll(async () => {
    hermes = await Hermes({
      durableName: "hermesTest",
      redisOptions: { ...redisConfig },
      poolOptions: { min: 0, max: 20 },
    }).connect();
  });

  it("should request a reply for a service", async () => {
    const topic = "say-hello";

    const requestSchema = z.object({ name: z.string() });
    const responseSchema = z.object({ message: z.string() });

    const service = await hermes.registerService(
      topic,
      requestSchema,
      responseSchema
    );

    const requestData: z.infer<typeof requestSchema> = { name: "Joe" };

    // Register the reply
    service.reply(async ({ reqData }) => {
      expect(reqData).toMatchObject(requestData);
      return { message: `Hello, ${reqData.name}!` };
    });

    const response = await service.request(requestData);

    expectTypeOf(response).toMatchTypeOf<z.infer<typeof responseSchema>>();
    expect(response).toMatchSnapshot();
  });

  afterAll(async () => {
    await hermes.disconnect();
  });
});

afterAll(async () => {
  testRedis.disconnect();
});
