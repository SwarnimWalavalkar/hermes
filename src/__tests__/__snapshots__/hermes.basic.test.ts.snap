// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Service > should request a reply for a service 1`] = `
{
  "message": "Hello, Joe!",
}
`;
